src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -153,17 +180,134 @@ double GAN<Model, InitializationRuleType, Noise>::Evaluate(
+      predictors.n_rows, batchSize, false, false);
+  currentTarget = arma::mat(responses.memptr() + i, 1, batchSize, false,
+      false);
+
+  discriminator.Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+      outputParameterVisitor,
+      discriminator.network.back())), std::move(currentTarget));
+
+  noise.imbue( [&]() { return noiseFunction();} );
+  generator.Forward(std::move(noise));
+
+  discriminator.predictors.cols(numFunctions, numFunctions + batchSize - 1) =
+      boost::apply_visitor(outputParameterVisitor, generator.network.back());
+  discriminator.Forward(std::move(discriminator.predictors.cols(numFunctions,
+      numFunctions + batchSize - 1)));
+  discriminator.responses.cols(numFunctions, numFunctions + batchSize - 1) =
+      -arma::ones(1, batchSize);
+
+  currentTarget = arma::mat(discriminator.responses.memptr() + numFunctions,
+      1, batchSize, false, false);
+  res += discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+      outputParameterVisitor,
+      discriminator.network.back())), std::move(currentTarget));
+
+  return res;
+}
+
+template<
+  typename Model,
+  typename InitializationRuleType,
+  typename Noise,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, WGANGP>::value,
+                        double>::type
+GAN<Model, InitializationRuleType, Noise, PolicyType>::Evaluate(
+    const arma::mat& /* parameters */,
+    const size_t i,
+    const size_t /* batchSize */)
+{
+  if (!reset)
+    Reset();
+
+  currentInput = arma::mat(predictors.memptr() + (i * predictors.n_rows),
+      predictors.n_rows, batchSize, false, false);
+  currentTarget = arma::mat(responses.memptr() + i, 1, batchSize, false,
+      false);
+
+  discriminator.Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+      outputParameterVisitor,
+      discriminator.network.back())), std::move(currentTarget));
+
+  noise.imbue( [&]() { return noiseFunction();} );
+  generator.Forward(std::move(noise));
+
+  arma::mat generatedData = boost::apply_visitor(outputParameterVisitor,
+      generator.network.back());
+  discriminator.predictors.cols(numFunctions, numFunctions + batchSize - 1) =
+      generatedData;
+  discriminator.Forward(std::move(discriminator.predictors.cols(numFunctions,
+      numFunctions + batchSize - 1)));
+  discriminator.responses.cols(numFunctions, numFunctions + batchSize - 1) =
+      -arma::ones(1, batchSize);
+
+  currentTarget = arma::mat(discriminator.responses.memptr() + numFunctions,
+      1, batchSize, false, false);
+  res += discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+      outputParameterVisitor,
+      discriminator.network.back())), std::move(currentTarget));
+
+  // Gradient Penalty is calculated here.
+  double epsilon = math::Random();
```

Comment:
Should we use a static parameter if we aren't training?

Reply:
I don't think it's necessary, but I don't mind adding it either, your call.

Reply:
I'm not too sure either, I think at least it will not decrease the performance we currently see, so if you see a simple solution, I would go for it.

Reply:
Since the original paper doesn't ask for that, nor have I seen any implementations for it, I'd prefer if we skip.

Reply:
Fine for me.